src/main/java/com/aws/greengrass/deployment/activator/KernelUpdateActivator.java

```diff
@@ -119,7 +120,10 @@ void rollback(Deployment deployment, Throwable failureCause) {
         deployment.setErrorTypes(errorReport.getRight());
         deployment.setStageDetails(Utils.generateFailureMessage(failureCause));
 
-        deployment.setDeploymentStage(KERNEL_ROLLBACK);
+        final boolean bootstrapOnRollbackRequired = kernelAlternatives.prepareBootstrapOnRollbackIfNeeded(
+                kernel.getContext(), deploymentDirectoryManager, bootstrapManager);
+
+        deployment.setDeploymentStage(bootstrapOnRollbackRequired ? ROLLBACK_BOOTSTRAP : KERNEL_ROLLBACK);
```

Comment:
how does this work if we're rolling back to a nucleus which does not understand "ROLLBACK_BOOSTRAP"?
If you're upgrading to 2.12.0 and that fails, then it will rollback to 2.11.3, won't that fail to read this deployment stage?

Reply:
In the scenario you describe, at kernel launch, the 2.11.3 version of Nucleus will attempt to determine the deployment stage by invoking the method `kernelAlts.determineDeploymentStage`, which will return `KERNEL_ROLLBACK`. Bootstrap-on-rollback logic will not execute (and isn't present, even if one wanted to it to execute).

Counter-intuitively, the value `ROLLBACK_BOOTSTRAP` that we provide to `setDeploymentStage` here is only useful when bootstrap-on-rollback fails. Then, when the kernel subsequently launches with deploymentStage==DEFAULT, `KernelUpdateDeploymentTask` can examine the value and understand that bootstrap-on-rollback failed during one of the bootstrap steps, and then provide the appropriate deployment status (`FAILED_UNABLE_TO_ROLLBACK`).